extras/FvwmConsole/FvwmConsole.1
.\" t
.\" @(#)FvwmConsole.1x	7/16/96
.TH FvwmConsole 1x "July 16 1996"
.UC

.SH NAME
FvwmConsole \- the FVWM command input interface
.SH SYNOPSIS
FvwmConsole is spawned by fvwm, so no command line invocation will work.
From within the .fvwm2rc file or from a menu, FvwmConsole takes any
options xterm will take, except -name, and -title.

.RS
Module FvwmConsole -g 40x10 -fg black -bg green3
.RE

 The option -e can be used to choose its front-end:

.RS
FvwmConsole -e /usr/X11/lib/fvwm2/FvwmConsoleC.pl
.RE

Also X resources can be set in .Xdefaults file

.RS
FvwmConsole*VT100*geometry: 40x4
.br
FvwmConsole*font: 7x14
.RE

.SH DESCRIPTION
FvwmConsole allows the user to type fvwm commands into xterm, and
have them executed immediately. These commands are usually specified
in the .fvwm2rc file, or are bound to menu/mouse items as specified
in that file. This tools is particularly useful for testing new
configuration ideas, or for implementing temporary changes to your
environment.


.SH Command Editor

There are a few options. If GNU readline library is installed,
add '#define HAVE_READLINE' in Imakefile and compile. 
.br
If Perl5 is installed, FvwmConsoleC.pl can be used as a command editor.
This can be accomplished by either copying FvwmConsoleC.pl 
to fvwmlib directory as FvwmConsoleC or
invoking FvwmConsole with -e option.
For example:

.br
FvwmConsole -e /usr/lib/X11/fvwm2/FvwmConsoleC.pl
.br

If neither one is installed, command line editor will call 
gets function, which doesn't have editing functions.
.br
GNU readline and FvwmConsoleC.pl have some frequent used commands
in common as default.
These commands are similar to emacs. 
For more details, refer GNU readline man and info pages, and
FvwmConsoleC.pl man pages.

.RS
.PD 0
.TP 8
Cntl-A
- beginning of line
.TP
Cntl-B
- previous char
.TP
Cntl-D
- delete char
.TP
Cntl-E
- end of line
.TP
Cntl-F
- next char
.TP
Cntl-H
- backspace
.TP 
Cntl-K
- erase to the end of line
.TP
Cntl-N
- next line
.TP
Cntl-P
- previous line
.TP
Cntl-R
- search reverse
.TP
Cntl-U
- delete line
.TP
Meta-B
- previous word
.TP
Meta-F
- next word
.TP
Esc <
- beginning of history
.TP
Esc >
- end of history
.RE
.PD
.SH COPYRIGHTS
Copyright 1996, Toshi Isogai. No guarantees or warranties are provided.
Use this program at your own risk. Permission to use this program for
any purpose is given, as long as the copyright is kept intact. 


.SH INVOCATION
FvwmConsole can be invoked by inserting the line 'Module FvwmConsole' in
the .fvwm2rc file. This can be placed on a line by itself, if FvwmConsole
is to be spawned during fvwm's initialization, or can be bound to a
menu or mouse button or keystroke to invoke it later. Fvwm will search
directory specified in the ModulePath configuration option to attempt
to locate FvwmConsole.

.SH CONFIGURATION OPTIONS
FvwmConsole uses xterm. All resources set for xterm are inherited
unless overridden by command line options.

.SH SEE ALSO
xterm(1), FvwmConsoleC.pl(1x), GNU readline library

.SH AUTHOR
Toshi Isogai  


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

extras/fvwmperl/FvwmConsole.pl
#!/usr/local/bin/tkperl
# console.pl - combination FvwmTalk/FvwmDebug using fvwmmod.pl
# (c)1996 Dan Astoorian <djast@utopia.csas.com>
# Permission is hereby granted to freely distribute fvwmperl
# without restriction as long as the copyright and this copyright
# notice remain intact.


require 'fvwmmod.pl';

($winid, $contxt, @CmdArgs) = &InitModule;

require 'getopts.pl';
{
    local (@ARGV) = @CmdArgs;
    &Getopts('p');
}

@history = ();
# Initial 500 line limit (arbitrary)
$epacketlimit = $packetlimit = 500;
$defFont = '*helvetica*m*o*n*14*';

use Tk;

$MW = MainWindow->new;

$cmd = $MW->Entry(
    -width   => 80,
);
$cmd->bind('<Return>' => sub {
    $line = $cmd->get;
    unshift(@history, $line);
    pop(@history) if (@history > 20);
    &SendInfo(0, $line);
    $cmd->delete(0,end);
    $histindx = @history;
});

$tf = $MW->Frame();
#$tf->pack(-fill => 'x', -before => $pf);
$tf->pack(-side => 'top', -fill => 'x');

# Extras for packet tracing
if (defined($opt_p)) {
    ($pf, $packetlog, $pl) = &ScrollBox($MW, "Packets:", 'fixed', 20, 'yes');

    $buflbl = $tf->Label(
	-text => 'Packet buffer limit (lines):',
	-font => $defFont,
    );

    $bufentry = $tf->Entry(
	-width => 5,
	-textvariable => \$epacketlimit,
	-justify => right,
    );
    $bufentry->bind('<Return>' => sub {
	local($size, $lines);
	if ($epacketlimit > 0) {
	    $packetlimit = $epacketlimit;
	}
	$bufentry->delete(0,end);
	$bufentry->insert(0,$packetlimit);
	$size = $packetlog->index(end) - 1;
	$lines = ($size - $packetlimit);
	$packetlog->delete(0, $lines) if ($size > $packetlimit);
	if ($packetlimit < 20) {
	    $packetlog->configure(-height => $packetlimit);
	} else {
	    $packetlog->configure(-height => 20);
	}
    });


    $buflbl->pack(-side => 'left');
    $bufentry->pack(-side => 'left', -anchor => 'e');
    $bufentry->bind('<Control-u>' => sub {$bufentry->delete(0,end)});

    $tf->bind('<Enter>' => sub {$bufentry->focus});

    &AddHandler($M_NEW_PAGE, 'ListNewPage');
    &AddHandler($M_NEW_DESK, 'ListNewDesk');
    &AddHandler($M_ADD_WINDOW | $M_CONFIGURE_WINDOW, 'ListConfigWin');
    &AddHandler($M_LOWER_WINDOW | $M_RAISE_WINDOW | $M_DESTROY_WINDOW |
	$M_DEICONIFY | $M_MAP,
	'ListWindow');
    &AddHandler($M_WINDOWSHADE | $M_DEWINDOWSHADE, 'ListWindowShade');
    &AddHandler($M_FOCUS_CHANGE, 'ListFocus');
    &AddHandler($M_ICONIFY | $M_ICON_LOCATION, 'ListIcon');
    &AddHandler($M_WINDOW_NAME | $M_ICON_NAME | $M_RES_CLASS | $M_RES_NAME,
	'ListName');
    &AddHandler($M_CONFIG_INFO | $M_ERROR | $M_STRING, 'ListConfigInfo');
    &AddHandler($M_END_WINDOWLIST | $M_END_CONFIG_INFO, 'ListEndConfigInfo');
    &AddHandler($M_ICON_FILE | $M_DEFAULTICON | $M_MINI_ICON, 'ListUnknown');
}

&AddHandler($M_ERROR | $M_STRING, 'handleError');

$cmd->bind('<Control-p>' => \&histprev);
$cmd->bind('<Up>' => \&histprev);
$cmd->bind('<Control-n>' => \&histnext);
$cmd->bind('<Down>' => \&histnext);

$cmd->bind('<Control-c>' => \&clearErrors);

($ef, $errlog, $el) = &ScrollBox($MW, "Errors:", 'variable', 5, 'no');

$quitbut = $tf->Button(
    -text => "Quit",
    -font => $defFont,
    -foreground => 'red',
    -command => sub { &EndModule ; exit },
);
$quitbut->pack(-side => right, -anchor => 'e');

$clearbut = $tf->Button(
    -text => "Clear",
    -font => $defFont,
    -foreground => 'red',
    -command => sub { &clearErrors },
);
$clearbut->pack(-side => right, -anchor => 'e');

$lbl = $MW->Label(
    -font => $defFont,
    -foreground => 'blue',
    -text => "Commands:",
);
$lbl->pack(-anchor => 'sw');

sub histprev {
    $cmd->delete(0,end);
    $histindx = 0 if (++$histindx > @history);
    $cmd->insert(0, "$history[$histindx]");
};
sub histnext {
    $cmd->delete(0,end);
    $histindx = @history if (--$histindx < 0);
    $cmd->insert(0, "$history[$histindx]") if ($histindx >= 0);
};

$cmd->bind('<Control-u>' => sub {$cmd->delete(0,end)});
$cmd->pack(-fill => 'x');

$MW->focusFollowsMouse;
$MW->title("TkPerlConsole.pl");

&EventLoop;

&EndModule;
exit;

########################################################################
# Subroutines
########################################################################

sub ScrollBox {
    local ($parent, $label, $fontname, $initheight, $expand) = @_;

    $frame = $parent->Frame();
    $frame->bind('<Enter>' => sub {$cmd->focus});
    $frame->pack(-side => 'top', -expand => $expand, -fill => 'both');
    $lbl = $frame->Label(
	-font => $defFont,
	-foreground => 'blue',
	-text => $label,
    );
    $lbl->pack(-anchor => 'sw');
    $sbar = $frame->Scrollbar;
    $sbar->pack(-side => 'right', -fill => 'y');
    $lbox = $frame->Listbox(
	-yscrollcommand => [$sbar => 'set'],
	-font => $fontname,
	-height => $initheight,
	-selectmode => 'single',
    );
    $sbar->configure(-command => [$lbox => 'yview']);
    $lbox->pack(-expand => $expand, -fill => 'both',);
    ($frame,$lbox,$lbl,$sbar);
}

# tohex - convert a list of values from decimal to hex
sub tohex {
    foreach (@_) {
	$_ = sprintf("%lx", $_);
	tr/a-z/A-Z/;
    }
}

sub Quit {
    &showline( "$0 exiting\n");
    &EndModule;
    exit;
}

sub ListNewPage { 
    local($type, $x, $y, $desk) = @_;
    &showline( "new page\n     x $x\n     y $y\n     desk $desk\n");
    1;
}
sub ListNewDesk {
    local($type, $desk) = @_;
    &showline( "new desk\n     desk $desk\n");
    1;
}

sub ListConfigWin { 
    local($type, $id, $fid, $ptr, $x, $y, $w, $h,
	$desk, $flags, $th, $bw, $wbw, $wbh,
	$wrwi, $wrhi, $minw, $minh, $maxw, $maxh,
	$lblid, $pmid, $grav, $tc, $bc) = @_;
    $stype = "Add Window" if ($type == $M_ADD_WINDOW);
    $stype = "Config Window" if ($type == $M_CONFIGURE_WINDOW);
    &tohex($id, $fid, $ptr, $flags, $lblid, $pmid, $grav, $tc, $bc);
    &showline( <<"END");
$stype
     ID $id
     frame ID $fid
     fvwm ptr $ptr
     frame x $x
     frame y $y
     frame w $w
     frame h $h
     desk $desk
     flags $flags
     title height $th
     border width $bw
     window base width $wbw
     window base height $wbh
     window resize width increment $wrwi
     window resize height increment $wrhi
     window min width $minw
     window min height $minh
     window max width $maxw
     window max height $maxh
     icon label window $lblid
     icon pixmap window $pmid
     window gravity $grav
     text color pixel value $tc
     border color pixel value $bc
END
    1;
}

sub ListWindow {
    ($type, $id, $fid, $ptr) = @_;
    $stype = "raise" if ($type == $M_RAISE_WINDOW);
    $stype = "lower" if ($type == $M_LOWER_WINDOW);
    $stype = "destroy" if ($type == $M_DESTROY_WINDOW);
    $stype = "map" if ($type == $M_MAP);
    $stype = "de-iconify" if ($type == $M_DEICONIFY);
    &tohex($id, $fid, $ptr);

    &showline( "$stype\n     ID $id\n     frame ID $fid\n     fvwm ptr $ptr\n");
    1;
}

sub ListFocus { 
    ($type, $id, $fid, $ptr, $tc, $bc) = @_;
    &tohex($id, $fix, $ptr);
    &showline( "focus\n     ID $id\n     frame ID $fid\n     fvwm ptr $ptr\n");
    &showline( "     text color pixel value $tc\n     border color pixel value $bc\n");
    1;
}

sub ListWindowShade { 
    ($type, $id) = @_;
    &tohex($id);
    $stype = "???";
    $stype = "windowshade" if ($type == $M_WINDOWSHADE);
    $stype = "de-windowshade" if ($type == $M_DEWINDOWSHADE);
    &showline( "$stype\n     ID $id\n");
    1;
}

sub ListIcon {
    ($type, $id, $fid, $ptr, $x, $y, $w, $h) = @_;
    $stype = "iconify" if ($type == $M_ICONIFY);
    $stype = "icon location" if ($type == $M_ICON_LOCATION);
    &tohex($id, $fix, $ptr);
    &showline( "$stype\n     ID $id\n     frame ID $fid\n     fvwm ptr $ptr\n");
    &showline( "     icon x $x\n     icon y $y\n     icon w $w\n     icon h $h\n");
    1;
}

sub ListName {
    ($type, $id, $fid, $ptr, $value) = @_;
    $stype = "window name" if ($type == $M_WINDOW_NAME);
    $stype = "icon name" if ($type == $M_ICON_NAME);
    $stype = "window class" if ($type == $M_RES_CLASS);
    $stype = "class resource name" if ($type == $M_RES_NAME);
    &tohex($id, $fid, $ptr);
    &showline( "$stype\n     ID $id\n     frame ID $fid\n     fvwm ptr $ptr\n");
    &showline( "     $stype $value\n");
    1;
}

sub ListConfigInfo {
    ($type, $txt) = @_;
    $stype = "config_info" if ($type == $M_CONFIG_INFO);
    $stype = "error" if ($type == $M_ERROR);
    $stype = "string" if ($type == $M_STRING);
    &showline( "$stype\n    $txt\n");
    1;
}

sub ListEndConfigInfo {
    &showline( "end_config_info\n");
    1;
}

sub ListUnknown {
    &showline( "miscellaneous packet\n    $type\n");
    1;
}

sub clearErrors {
    $errlog->delete(0, end);
    $packetlog->delete(0, end) if (defined($opt_p));
}

sub handleError {
    $errlimit = 20;
    local($size, $lines);

    $errlog->insert(end,"$_[1]\n");
    $size = $errlog->index(end) - 1;
    $lines = ($size - $errlimit);
    $errlog->delete(0, $lines) if ($size > $errlimit);
    $errlog->see(end);
    1;
};

sub showline {
    local($size, $lines);

    for $line (split(/\n/,$_[0])) {
	$packetlog->insert(end,"$line");
    }
    $size = $packetlog->index(end) - 1;
    $lines = ($size - $packetlimit);
    $packetlog->delete(0, $lines) if ($size > $packetlimit);
    $packetlog->see(end);
}
